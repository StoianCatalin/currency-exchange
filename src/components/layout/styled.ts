import styled from 'styled-components';
import {
    paddingSize, white, baseBorders, marginSize, grayColor,
    device, size, primaryColor, borderRadiusSize, primaryColorOnHover, darkColor
} from 'styles';

export const Layout = styled.div`
    width: 100%;
    margin: auto;
`;

export const Card = styled.div`
    box-sizing: border-box;
    background-color: ${white};
    padding: ${paddingSize};
    height: 100vh;

    @media ${device.tablet} {
        ${baseBorders}
        max-width: ${size.tablet};
        margin: ${marginSize} auto;
        height: auto;
    }
`;

export const Divider = styled.div`
    margin: 15px 0;
    width: 100%;
    height: 1px;
    background-color: ${grayColor};
`;

export const Button = styled.button`
    width: 100%;
    background-color: ${primaryColor};
    color: ${white};
    border: 0;
    height: 40px;
    border-radius: ${borderRadiusSize};
    font-weight: bold;
    cursor: pointer;

    &:hover {
        background-color: ${primaryColorOnHover};
    }
`;

export const Text = styled.span`
    font-size: ${(props: { size: number }) => `${props.size}px`};
    color: ${darkColor};
`;
