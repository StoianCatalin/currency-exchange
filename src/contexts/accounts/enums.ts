
export enum CURRENCY_SYMBOLS { 
    EUR = "eur",
    USD = "usd",
    GBP = "gbp",
    RON = "ron"
 }

export enum INPUT_TYPES {
    FROM = "from",
    TO = "to"
}