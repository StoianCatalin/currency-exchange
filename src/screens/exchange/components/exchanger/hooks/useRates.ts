import { useEffect, useRef, useState } from "react";
import ExchangeRateService from "services/exchange-rate";
import { ExchangeRateResponse } from "services/exchange-rate/types";

interface IRates {
    rates: ExchangeRateResponse["conversion_rates"];
    isLoading: boolean;
    setCurrency: (currency: string) => void;
}
const TIME_TO_REFRESH = 10000; // 10 seconds;

export default function useRates(token: string, refreshTime = TIME_TO_REFRESH): IRates {
    const [rates, setRates] = useState<ExchangeRateResponse["conversion_rates"]>({});
    const [currency, setCurrency] = useState<string>(token);
    const [isLoading, setIsLoading] = useState<boolean>(false);
    const timer = useRef<ReturnType<typeof setTimeout>|null>(null);
    const exchangeRateService = new ExchangeRateService();

    const fetchRates = async () => {
        setIsLoading(true);
        const { conversion_rates } = await exchangeRateService.getRate(currency);
        setRates(conversion_rates);
        setIsLoading(false);
    };

    // This will be executed each time rates is changed and will trigger a new timer for the next TIME_TO_REFRESH seconds.
    useEffect(() => {
        if (timer.current) {
            clearTimeout(timer.current);
        }
        timer.current = setTimeout(async () => {
            const { conversion_rates } = await exchangeRateService.getRate(currency);
            setRates(conversion_rates);
        }, refreshTime);

        return () => {
            if (timer.current) {
                clearTimeout(timer.current);
            }
        }
    }, [rates]);

    // This will be executed to updated the rates according with new from currency
    useEffect(() => {
        (async () => {
            await fetchRates();
        })();
    }, [currency]);

    return {
        rates,
        setCurrency,
        isLoading
    };
};
