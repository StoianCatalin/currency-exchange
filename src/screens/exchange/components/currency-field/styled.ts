import styled from 'styled-components';
import Select from 'react-select';
import { borderLargeRadiusSize, lightColor, paddingSize } from 'styles';

export const InputWrapper = styled.div`
    width: 100%;
    border-radius: ${borderLargeRadiusSize};
    min-height: 100px;
    background-color: ${lightColor};
    margin: 10px 0;
    padding: ${paddingSize};
    box-sizing: border-box;
    display: flex;
`;

export const Column = styled.div`
    width: 50%;
`;

export const SelectDropdown = styled(Select)`
    width: 90%;
`;
