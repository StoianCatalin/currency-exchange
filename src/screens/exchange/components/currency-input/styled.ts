import styled from 'styled-components';
import { Text } from 'components/layout';
import { titleSize, errorColor } from 'styles';

export const Wrapper = styled.div`
    width: 100%;
    display: flex;
    flex-direction: column;
`;

export const Input = styled.input`
    width: 100%;
    height: 30px;
    border: 0;
    background-color: transparent;
    font-size: ${titleSize};
    font-weight: bold;
    text-align: right;

    &:focus {
        border: none;
        outline: none;
    }
`;

export const ErrorText = styled(Text)`
    color: ${errorColor};
    text-align: right;
`;
