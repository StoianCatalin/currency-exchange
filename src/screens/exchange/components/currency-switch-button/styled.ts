import styled from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { darkColor, primaryColor, secondaryColor, white } from 'styles';

export const SwitchButton = styled.div`
    width: 30px;
    height: 30px;
    border-radius: 50%;
    margin: auto;
    cursor: pointer;
    background-color: ${primaryColor};
    color: ${white};
    display: flex;
    justify-content: center;
    align-items: center;
    transition: ease .2s background-color;

    &:hover {
        color: ${darkColor};
        background-color: ${secondaryColor}
    }
`;

export const StyledIcon = styled(FontAwesomeIcon)`
    transform: rotate(-90deg);
`;