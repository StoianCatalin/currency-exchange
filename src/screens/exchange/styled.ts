import styled from 'styled-components';
import { primaryColor, textSize, titleSize } from 'styles';


export const Title = styled.h2`
    font-size: ${titleSize};
    margin: 0;
`;

export const Subtitle = styled.span`
    color: ${primaryColor};
    font-size: ${textSize};
    display: flex;
    gap: 6px;
    align-items: center;
`;
