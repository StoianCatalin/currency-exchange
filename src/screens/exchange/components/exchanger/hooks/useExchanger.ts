import { AccountContext } from "contexts/accounts";
import { AddCurrencyAction, SubtractCurrencyAction } from "contexts/accounts/actions";
import { CURRENCY_SYMBOLS, INPUT_TYPES } from "contexts/accounts/enums";
import ICurrency from "interfaces/currency";
import React, { useContext, useState } from "react";
import { ExchangeRateResponse } from "services/exchange-rate/types";
import calculateValueBasedOnRate from "../utils/calculateValueBasedOnRate";
import useRates from "./useRates";

interface IInputState {
    currency: CURRENCY_SYMBOLS;
    amount: string;
}

export interface IExchanger {
    fromInput: IInputState;
    toInput: IInputState;
    rates: ExchangeRateResponse["conversion_rates"];
    onChangeCurrency: (newCurrency: ICurrency, type: INPUT_TYPES) => void;
    switchCurrencies: () => void;
    makeTransaction: () => void;
    onInputChange: (type: INPUT_TYPES, value: string) => void;
}

export default function useExchanger(): IExchanger {
    const { state, dispatch } = useContext(AccountContext);
    const [fromInput, setFromInput] = useState({ currency: CURRENCY_SYMBOLS.EUR, amount: "0" });
    const [toInput, setToInput] = useState({ currency: CURRENCY_SYMBOLS.USD, amount: "0" });
    const { rates, setCurrency } = useRates(fromInput.currency);

    /*
    Function used to change the currency for a input in exchanger.
    * @param {ICurrency} newCurrency
    * @param {"from"|"to"} type
    */
    function onChangeCurrency(newCurrency: ICurrency, type: INPUT_TYPES) {
        switch(type) {
            case INPUT_TYPES.FROM:
                if (newCurrency.value === toInput.currency) {
                    setToInput({ ...toInput, currency: fromInput.currency });
                }
                setFromInput({ ...fromInput, currency: newCurrency.value});
                setCurrency(newCurrency.value);
                break;
            case INPUT_TYPES.TO:
                if (newCurrency.value === fromInput.currency) {
                    setFromInput({ ...toInput, currency: toInput.currency });
                }
                setToInput({ amount: calculateValueBasedOnRate(rates, fromInput.amount, newCurrency.value), currency: newCurrency.value});
                break;
            default:
                break;
        }
    };

    /*
    Function used to interchange the currencies between inputs.
    */
    function switchCurrencies() {
        setCurrency(toInput.currency);
        setFromInput({ ...toInput });
        setToInput({ ...fromInput });
    };

    /* 
    Function that will handle when one of the inputs are changed (From or To input)
    */
    function onInputChange(type: INPUT_TYPES, value: string) {
        switch(type) {
            case INPUT_TYPES.FROM:
                setFromInput({ ...fromInput, amount: value});
                setToInput({ ...toInput, amount: calculateValueBasedOnRate(rates, value, toInput.currency) });
                break;
            case INPUT_TYPES.TO:
                setFromInput({ ...fromInput, amount: calculateValueBasedOnRate(rates, value, toInput.currency, true) })
                setToInput({ ...toInput, amount: value});
                break;
            default:
                break;
        }
    }

    /* 
    Function used to validate and execute the transaction.
    */
    function makeTransaction() {
        if (parseFloat(fromInput.amount) > state[fromInput.currency]) {
            return;
        }
        if (fromInput.amount === "0" || toInput.amount === "0") {
            return;
        }
        dispatch(new SubtractCurrencyAction({
            currency: fromInput.currency,
            amount: parseFloat(fromInput.amount)
        }));
        dispatch(new AddCurrencyAction({
            currency: toInput.currency,
            amount: parseFloat(toInput.amount)
        }));
        resetInputs();
    }

    function resetInputs() {
        setFromInput({ ...fromInput, amount: "0" });
        setToInput({ ...toInput, amount: "0" });
    }

    return {
        fromInput,
        toInput,
        rates,
        onChangeCurrency,
        switchCurrencies,
        onInputChange,
        makeTransaction
    };
}
