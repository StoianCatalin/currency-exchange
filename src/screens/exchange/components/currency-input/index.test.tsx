import { CURRENCY_SYMBOLS, INPUT_TYPES } from 'contexts/accounts/enums';
import renderer from 'react-test-renderer';
import React from 'react';
import CurrencyInput from '.';
import { IExchanger } from '../exchanger/hooks/useExchanger';
import { ErrorText, Input } from './styled';


describe("CurrencyInput", () => {
    let exchangerMock: IExchanger;
    const fromInput = { currency: CURRENCY_SYMBOLS.EUR, amount: 0 };

    beforeEach(() => {
        exchangerMock = ({
            onInputChange: jest.fn().mockImplementation((type, value) => {
                fromInput.amount = value;
            }),
            fromInput
        }) as any;
    });

    afterAll(() => {
        jest.clearAllMocks();
    });


    it("should render correctly", () => {
        // When
        const component = renderer.create(
            <CurrencyInput
                exchanger={exchangerMock}
                type={INPUT_TYPES.FROM}
            />
        );
        // Then
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    describe("onChange", () => {
        it("should remove non-numeric characters", () => {
            const component = renderer.create(
                <CurrencyInput
                    exchanger={exchangerMock}
                    type={INPUT_TYPES.FROM}
                />
            );
            const input = component.root.findByType(Input);
            const event = { target: { value: '-20' } } as any;
            input.props.onChange(event);
    
            expect(exchangerMock.onInputChange).toBeCalledWith(INPUT_TYPES.FROM, "20");
        });

        it("should keep the dot (.) if it is the last character", () => {
            const component = renderer.create(
                <CurrencyInput
                    exchanger={exchangerMock}
                    type={INPUT_TYPES.FROM}
                />
            );
            const input = component.root.findByType(Input);
            const event = { target: { value: '-20.' } } as any;
            input.props.onChange(event);
    
            expect(exchangerMock.onInputChange).toBeCalledWith(INPUT_TYPES.FROM, "20.");
        });

        it("should don't allow for multiple dots", () => {
            const component = renderer.create(
                <CurrencyInput
                    exchanger={exchangerMock}
                    type={INPUT_TYPES.FROM}
                />
            );
            const input = component.root.findByType(Input);
            const event = { target: { value: '-20..' } } as any;
            input.props.onChange(event);
    
            expect(exchangerMock.onInputChange).toBeCalledWith(INPUT_TYPES.FROM, "20");
        });

        it("should return maximum 2 decimals after the .", () => {
            const component = renderer.create(
                <CurrencyInput
                    exchanger={exchangerMock}
                    type={INPUT_TYPES.FROM}
                />
            );
            const input = component.root.findByType(Input);
            const event = { target: { value: '-20.321' } } as any;
            input.props.onChange(event);
    
            expect(exchangerMock.onInputChange).not.toBeCalled();
        });
    });

    describe("isError", () => {
        it("should return false if amount is below the state value", () => {
            const component = renderer.create(
                <CurrencyInput
                    exchanger={exchangerMock}
                    type={INPUT_TYPES.FROM}
                />
            );
            expect(() => {
                component.root.findByType(ErrorText)
            }).toThrowError();
        });

        it("should return true if amount is above the state value", () => {
            exchangerMock.fromInput.amount = 210;
            const component = renderer.create(
                <CurrencyInput
                    exchanger={exchangerMock}
                    type={INPUT_TYPES.FROM}
                />
            );
            const element = component.root.findByType(ErrorText);
            expect(element).toBeTruthy();
        });
    });
});
