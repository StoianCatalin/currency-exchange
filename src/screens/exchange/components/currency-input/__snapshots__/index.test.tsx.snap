// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyInput should render correctly 1`] = `
<div
  className="sc-jSFjdj lmQA-DU"
>
  <input
    className="sc-gKAaRy ipBHhz"
    onChange={[Function]}
    placeholder="0.00"
    value="-0"
  />
</div>
`;
